[WIP] Keep connectionstring safe

melkio opened the pull request:
Connects to https://github.com/Particular/PlatformDevelopment/issues/718
(implements https://github.com/Particular/docs.particular.net/issues/1268)

SeanFeldman commented on nservicebus/azure/intro-keep-connection-string-safe.include.md:
```diff
@@ -0,0 +1,4 @@
+It is possible to accidentally leak sensitive information in the connection string if it's not properly secured. E.g. the information can be leaked if an error occurs when communicating across untrusted boundaries, or if the error information is logged to an unsecured log file.
+
+In order to prevent it, transport, versions 7 and above, allows for creating a logical name for each connection string. The name is mapped to the physical connection string, and connection strings are always referred to by their logical name.   
```
If the message is on the queue, what would `ReplyTo` header contain for example? Would it help to show the difference?

SeanFeldman commented on nservicebus/azure/intro-keep-connection-string-safe.include.md:
```diff
@@ -0,0 +1,4 @@
+It is possible to accidentally leak sensitive information in the connection string if it's not properly secured. E.g. the information can be leaked if an error occurs when communicating across untrusted boundaries, or if the error information is logged to an unsecured log file.
```
After reading this I'm still not sure how this affects me as an NSB user. Does it mean the log files are not sure? What are the boundaries that the reference made to?

melkio replied:
@SeanFeldman I only brought ASQ related section into an include file and, as suggested by @yvesgoeleven, linked it from a common place to avoid copy&paste and docu duplication.
I did nothing on this paragraphs, I'll work on it soon. I've read your comments about it somewhere ;)

SeanFeldman commented:
@melkio :+1: Upgrade guide would definitely need to have clear explanation what and why.

SeanFeldman commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,32 @@
+
+include: keep-connection-string-safe
+
+To enable this behavior, apply the following configuration:
```
Putting my non-maintainer :tophat: on. 

>To enable this behavior, apply the following configuration:

What is this behaviour? What does it do? Why would I wan to use it?
This is kind of information I'd like to find in the first paragraph and after that see _how_ to do it :)

SeanFeldman replied:
More info:

As a reader, I can be either a newbie, or someone who's already using nsb.asb and wants to read about this new features.

- For a new user there's no existing context (the fact that connection strings where not safe).
- For an existing user they know what's happening, but they don't know about how safe connection strings work 
  - link to the upgrade guide
- For both type of users, how to configure safe connections

Would this flow make sense?

SeanFeldman replied:
Overlooked the `keep-connection-string-safe` part.
Would it make sense to elaborate more and provide a little example (tiny snippet of an output) to demonstrate the problem?

SeanFeldman replied:
With snippet like that each time we add something, we have to keep in mind that it's going to be used in multiple places. Is that kind of re-use is desired? I don't think this "duplication" is bad. Plus, the ability of showing what would be rendered in version 6 w/o safe connection strings would be hurt since for ASB and ASQ it's not the same connection string type. My 2c :)

petersgiles replied:
I agree with all of @SeanFeldman's points

SeanFeldman commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,32 @@
+
+For detailed explanation about all supported partitioning strategies and how to configure namespaces mapping for them see related [article](multiple-namespaces-support.md).
+
+## How to upgrade endpoints
```
Should this be in the upgrade guide?

SeanFeldman commented on nservicebus/upgrades/asb-6to7.md:
```diff
@@ -59,4 +59,14 @@ snippet:topology-selection-upgrade-guide
+
+## Keep connection string safe
+
+In version 7 and above, to avoid sharing sensitive data contained into connection strings, it's possibile to enable `UseNamespaceNameInsteadOfConnectionString` feature: 
```
I'm confused a little. Upgrade guide is intended for thins that have changed between old and new version and **must** be done in order to work properly. If by default, connection strings are used, that means during upgrade there's nothing  one must do in order for solution to work. If enabling safe connection strings is optional, it shouldn't be in the upgrade guide. Please correct me if I'm wrong with assuming that by default we're backwards compatible and one has to opt in to enable safe connection strings. Thank you.

SeanFeldman commented on nservicebus/azure/keep-connection-string-safe.include.md:
```diff
@@ -0,0 +1,4 @@
+Transport, versions 6 and below, uses raw connection string provided into message headers. It could result in a leak of sensitive information, f.i. logging it.
```
Where's this include is used?

petersgiles commented:
Overall I think  there needs to be a better description of what the feature is, who would use it and when and how.

The snippets seem good.

I'm also a little hazy on what I am required to upgrade and what I don't have to worry about

melkio commented:
> What is this behaviour? What does it do? Why would I wan to use it?
This is kind of information I'd like to find in the first paragraph and after that see how to do it :)

@SeanFeldman the first paragraph is into the _include_ file. The "behavior" is mentioned there.

melkio commented:
@Particular/azure-service-bus-maintainers @petersgiles new version. I tried to address your comments. Let me know...

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+reviewed: 2016-04-26
+---
+
+Transport, versions 6 and below, uses raw connection string provided into message headers.  
```
how about 

> Versions 6 and below uses raw the connection string in message headers. For instance, a typical value for `ReplyTo` header could be:

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+
+`[queue name]@Endpoint=sb://[namespace name].servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[secrets here]`
+  
+It could result in a leak of sensitive information, f.i. logging it.  
```
`f.i.`?

SimonCropp replied:
how about 

> This could potentially result in a leak of sensitive information, for example in [log files](/nservicebus/logging/) or in the [error queue](/nservicebus/errors/).

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+`[queue name]@Endpoint=sb://[namespace name].servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[secrets here]`
+  
+It could result in a leak of sensitive information, f.i. logging it.  
+In order to prevent it, transport, versions 7 and above, introduces the needs to map logical namespace name to namespace connection string. This configuration allows transport to convert connection string value to namespace name, if mapping exists, for each incoming message, using the logical name internally.
```
how about 

> In order to prevent this Versions 7 and above includes the ability to map logical namespace name to namespace connection string.

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+  
+It could result in a leak of sensitive information, f.i. logging it.  
+In order to prevent it, transport, versions 7 and above, introduces the needs to map logical namespace name to namespace connection string. This configuration allows transport to convert connection string value to namespace name, if mapping exists, for each incoming message, using the logical name internally.
+If a valid mapping between connection string and namespace name doesn't exist, transport throws a `KeyNotFoundException`.
```
how about 

> If a valid mapping between connection string and namespace name doesn't exist a [`KeyNotFoundException`](https://msdn.microsoft.com/en-us/library/system.collections.generic.keynotfoundexception.aspx) is thrown.

SimonCropp replied:
also why KeyNotFoundException? should we give them an `Exception` with a sensible message including the key that was attempted?

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+In order to prevent it, transport, versions 7 and above, introduces the needs to map logical namespace name to namespace connection string. This configuration allows transport to convert connection string value to namespace name, if mapping exists, for each incoming message, using the logical name internally.
+If a valid mapping between connection string and namespace name doesn't exist, transport throws a `KeyNotFoundException`.
+
+To configure mapping between a namespace name and the corresponding connection string is possible through `AddNamespace(string name, string connectionString)` configuration API, as shown below for `SingleNamespacePartitioning` strategy:
```
how about 

> Mapping between a namespace name and the corresponding connection string is done as follows:

mentioning inline code when a snippet is below is kind of redundant

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+
+snippet: map_logical_name_to_connection_string
+
+For detailed explanation about all supported partitioning strategies and how to configure namespaces mapping for them see related [article](multiple-namespaces-support.md).  
```
how about 

> For detailed explanation about all supported partitioning strategies, and how to configure namespaces mapping, see [Multiple Namespaces Support](multiple-namespaces-support.md).

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+
+snippet: map_default_logical_name_to_connection_string 
+
+In the event of an error or when logging only, the logical name will be used avoiding sharing of sensitive information.
```
> In the event of an error or when logging only

this is ambiguous. "event of an error" sounds like u are referring to when an exception is thorn.

"when logging only" why "only"?

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+
+In the event of an error or when logging only, the logical name will be used avoiding sharing of sensitive information.
+
+To enable the same behavior also for outgoing messages, a new feature, provided by transport version 7 and above, has to be switched on:
```
How about

> To enable the same behavior also for outgoing messages, a new feature, provided in Version 7 and above, has to be enabled:

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+
+snippet: enable_use_namespace_name_instead_of_connection_string
+
+Without enabling this behavior, transport converts back the value to `queueName@connectionString` before delivering the message, to ensure backward compatibility among endpoints of different versions.  
```
i think this would be better done with two headings "versions 6 and below" and "versions 7 and above" and explain the behavior in each of those

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+Without enabling this behavior, transport converts back the value to `queueName@connectionString` before delivering the message, to ensure backward compatibility among endpoints of different versions.  
+Enabling `UseNamespaceNameInsteadOfConnectionString` feature, transport ensures the same behavior, avoiding sharing of sensitive information, for all incoming and outgoing messages.
+
+For explanation about how to upgrade endpoints to enable this feature, see [upgrade guide article](/nservicebus/upgrades/asb-6to7.md)
```
make this a related in the header instead 

```
related:
 - nservicebus/upgrades/asb-6to7
```

SimonCropp commented on nservicebus/upgrades/asb-6to7.md:
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
+- upgrade all endpoints to version 7 or above.
+- switch on `UseNamespaceNameInsteadOfConnectionString` feature one endpoint at time. 
+
+Refer to the [Keep connection string safe page](/nservicebus/azure-service-bus/keep-connectionstring-safe.md) for more details.
```
delete this and make the heading a link instead

SimonCropp commented on nservicebus/upgrades/asb-6to7.md:
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
+In order to be able to enable `UseNamespaceNameInsteadOfConnectionString` follow the next steps:
```
> In order to be able to enable

i think you need to re-word that

SimonCropp commented on nservicebus/upgrades/asb-6to7.md:
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
+
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
```
what does this have to do with "Keep connection string safe" ?

SimonCropp commented on nservicebus/upgrades/asb-6to7.md:
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
+In order to be able to enable `UseNamespaceNameInsteadOfConnectionString` follow the next steps:
+- upgrade all endpoints to version 7 or above.
+- switch on `UseNamespaceNameInsteadOfConnectionString` feature one endpoint at time. 
```
why "one endpoint at time"?

SimonCropp replied:
do people need to deploy all endpoints with v7 prior to enabling UseNamespaceNameInsteadOfConnectionString? if so that is not clear in the above

melkio replied:
@SimonCropp without deploying v7 transport hasn't `UseNamespaceNameInsteadOfConnectionString` option. Until you have a mixed configuration (v6 and v7 endpoints) this option can't be enabled. it's not backward compatible

I try to rephrase

SeanFeldman replied:
> I'll try to rephrase

:heart:

SimonCropp commented on nservicebus/azure-service-bus/keep-connectionstring-safe.md:
```diff
@@ -0,0 +1,40 @@
+---
```
should this be in the menu?

melkio commented:
@SimonCropp I addressed all your comments (thanks) except for this one:

> i think this would be better done with two headings "versions 6 and below" and "versions 7 and above" and explain the behavior in each of those

I don't get your point. Could you explain better? Because I don't know how to explain a behavior that doesn't exist on v6 and below :) The first paragraph explains how transport v6 works

SimonCropp commented:
i updated some minor things

SimonCropp commented:
i think prior to merging this we need to address this https://github.com/Particular/NServiceBus.AzureServiceBus/issues/217

SeanFeldman commented:
Question: is `asb-credential-warning.include.md` used elsewhere?

melkio commented:
@SimonCropp take a look to comments here: https://github.com/Particular/NServiceBus.AzureServiceBus/issues/217

SimonCropp commented:
merged manually here https://github.com/Particular/docs.particular.net/commit/9eb42314b19bb55fed835f9cde2b1bbe4b95abe7
